Tactics/SymContext.lean

```diff
@@ -0,0 +1,257 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Alex Keizer
+-/
+import Lean
+import Arm.Exec
+
+/-!
+This files defines the `SymContext` structure,
+which is collects
+-/
+
+open Lean Meta
+open BitVec
+
+/--
+
+-/
+structure SymContext where
+  /-- `state` and `finalState` are local variable of type `ArmState` -/
+  state : Name
+  -- TODO: Should we eventually track the final state as well?
+  --       We could use it to avoid introducing the last intermediate state,
+  --       when we detect that `runSteps = 0`
+  /-- `runSteps` is the number of steps that we can *maximally* simulate,
+  because of the way it occurs in `h_run`.
+  Note that `runSteps` is a meta-level natural number, reflecting the fact that we expect the number
+  of steps in `h_run` to be expressed as a concrete literal -/
+  runSteps : Nat
+  /-- `h_run` is a local hypothesis of the form `finalState = run {runSteps} state` -/
+  h_run : Name
+  /-- `program` is a *constant* which represents the program being evaluated -/
+  program : Name
+  /-- `h_program` is a local hypothesis of the form `state.program = program` -/
+  h_program : Name
+  /-- `pc` is the *concrete* value of the program counter -/
+  pc : BitVec 64
+  /-- `h_pc` is a local hypothesis of the form `r StateField.PC state = pc` -/
+  h_pc  : Name
+  /-- `h_err` is a local hypothesis of the form `r StateField.ERR state = .None` -/
+  h_err : Option Name
+  /-- `h_sp` is a local hypothesis of the form `CheckSPAlignment state` -/
+  h_sp  : Option Name
+
+  /-- `state_prefix` is used together with `curr_state_number`
+  to determine the name of the next state variable that is added by `sym` -/
+  state_prefix      : String := "s"
+  /-- `curr_state_number` is incremented each simulation step,
+  and used together with `curr_state_number`
+  to determine the name of the next state variable that is added by `sym`  -/
+  curr_state_number : Nat := 0
+
+
+namespace SymContext
+
+/-! ## Creating initial contexts -/
+
+/-- Infer `state_prefix` and `curr_state_number` from the `state` name as follows:
+if `state` is `s{i}` for some number `i` and a single character `s`,
+then `s` is the prefix and `i` the number,
+otherwise simple take `{state}_` as the prefix and start counting from zero -/
+def inferStatePrefixAndNumber (ctxt : SymContext) : SymContext :=
+  let state := ctxt.state.toString
+  let tail := state.toSubstring.drop 1
+
+  if let some curr_state_number := tail.toNat? then
+    { ctxt with
+      state_prefix := (state.get? 0).getD 's' |>.toString,
+      curr_state_number }
+  else
+    { ctxt with
+      state_prefix := state ++ "_",
```

Comment:
Crucially, the underscore is added in the `else` branch, so when the state variable *isn't* a single character followed by a number, e.g., if the initial state is called `initialState`, or `state`.

Admittedly, I didn't give it that much thought, this would indeed behave badly if there are multiple `sym_n` calls.
I've refactored it to ignore the `state` variables name if it can't be decomposed, and start counting from `s1` (assuming the initial state should've been `s0`) --- the alternative would be some logic that tries to split on the last underscore and checks if the rest is a `Nat` to ensure round-tripping, but that seemed too complex.